File: backend/app/schemas/legal_consent.py
from datetime import datetime

from pydantic import BaseModel, Field


CONSENT_TYPES = ("terms", "privacy", "biometric")
LEGAL_TYPES = ("terms", "privacy")


class ConsentTypeStatus(BaseModel):
    consent_type: str
    published: bool
    satisfied: bool
    current_version: str | None = None
    accepted_version: str | None = None
    accepted_at: datetime | None = None
    last_action: str | None = None
    content_source: str | None = None


class LegalConsentStatusResponse(BaseModel):
    business_name: str
    business_code: str
    legal_page_url: str
    legal_page_api_path: str
    is_demo: bool = False
    use_custom_consents: bool = False
    legal_satisfied: bool
    biometric_satisfied: bool
    consents_completed: bool = False
    adult_acknowledged: bool = False
    terms: ConsentTypeStatus
    privacy: ConsentTypeStatus
    biometric: ConsentTypeStatus


class LegalConsentAcceptRequest(BaseModel):
    accepted: bool = Field(description="Must be true to record consent")
    types: list[str] = Field(min_length=1)
    versions: dict[str, str]
    client: str = Field(default="mobile")
    adult_acknowledged: bool = False


class LegalConsentWithdrawRequest(BaseModel):
    client: str = Field(default="web")


class PublicLegalSection(BaseModel):
    title: str
    consent_type: str
    content: str | None = None
    version: str | None = None
    published: bool
    content_source: str | None = None


class PublicLegalPageResponse(BaseModel):
    business_name: str
    business_code: str
    legal_page_url: str
    updated_at: datetime | None = None
    is_demo: bool = False
    use_custom_consents: bool = False
    terms: PublicLegalSection
    privacy: PublicLegalSection
    biometric: PublicLegalSection


class PlatformLegalSection(BaseModel):
    title: str
    consent_type: str
    content: str | None = None
    version: str | None = None
    published: bool
    updated_at: datetime | None = None
    updated_by: str | None = None


class PlatformLegalSettingsResponse(BaseModel):
    terms: PlatformLegalSection
    privacy: PlatformLegalSection
    biometric: PlatformLegalSection


class PlatformLegalSettingsUpdate(BaseModel):
    terms_content: str | None = None
    privacy_content: str | None = None
    biometric_consent_content: str | None = None
    terms_version: str | None = Field(default=None, max_length=80)
    privacy_version: str | None = Field(default=None, max_length=80)
    biometric_consent_version: str | None = Field(default=None, max_length=80)
    terms_published: bool | None = None
    privacy_published: bool | None = None
    biometric_published: bool | None = None


class EffectiveLegalBundle(BaseModel):
    updated_at: datetime | None = None
    terms: PublicLegalSection
    privacy: PublicLegalSection
    biometric: PublicLegalSection


class CustomLegalDraft(BaseModel):
    terms_content: str | None = None
    privacy_content: str | None = None
    biometric_consent_content: str | None = None
    terms_version: str | None = None
    privacy_version: str | None = None
    biometric_consent_version: str | None = None
    legal_updated_at: datetime | None = None
    terms_published: bool
    privacy_published: bool
    biometric_published: bool


class BusinessLegalSettingsResponse(BaseModel):
    business_name: str
    business_code: str
    legal_page_url: str
    use_custom_consents: bool
    effective: EffectiveLegalBundle
    defaults: PlatformLegalSettingsResponse
    custom: CustomLegalDraft


class BusinessLegalSettingsUpdate(BaseModel):
    use_custom_consents: bool | None = None
    terms_content: str | None = None
    privacy_content: str | None = None
    biometric_consent_content: str | None = None
    terms_version: str | None = Field(default=None, max_length=80)
    privacy_version: str | None = Field(default=None, max_length=80)
    biometric_consent_version: str | None = Field(default=None, max_length=80)


class EmployeeConsentSummary(BaseModel):
    terms_satisfied: bool = False
    privacy_satisfied: bool = False
    biometric_satisfied: bool = False
    terms_version: str | None = None
    privacy_version: str | None = None
    biometric_version: str | None = None
    terms_accepted_at: datetime | None = None
    privacy_accepted_at: datetime | None = None
    biometric_accepted_at: datetime | None = None

File: mobile/lib/domain/entities/legal_consent.dart
class LegalSection {
  const LegalSection({
    required this.title,
    required this.consentType,
    required this.published,
    this.content,
    this.version,
    this.contentSource,
  });

  final String title;
  final String consentType;
  final bool published;
  final String? content;
  final String? version;
  final String? contentSource;

  factory LegalSection.fromJson(Map<String, dynamic> json) {
    return LegalSection(
      title: json['title'] as String? ?? '',
      consentType: json['consent_type'] as String? ?? '',
      published: json['published'] == true,
      content: json['content'] as String?,
      version: json['version'] as String?,
      contentSource: json['content_source'] as String?,
    );
  }
}

class PublicLegalPage {
  const PublicLegalPage({
    required this.businessName,
    required this.businessCode,
    required this.legalPageUrl,
    required this.terms,
    required this.privacy,
    required this.biometric,
  });

  final String businessName;
  final String businessCode;
  final String legalPageUrl;
  final LegalSection terms;
  final LegalSection privacy;
  final LegalSection biometric;

  factory PublicLegalPage.fromJson(Map<String, dynamic> json) {
    return PublicLegalPage(
      businessName: json['business_name'] as String? ?? '',
      businessCode: json['business_code'] as String? ?? '',
      legalPageUrl: json['legal_page_url'] as String? ?? '',
      terms: LegalSection.fromJson(
        json['terms'] as Map<String, dynamic>? ?? const {},
      ),
      privacy: LegalSection.fromJson(
        json['privacy'] as Map<String, dynamic>? ?? const {},
      ),
      biometric: LegalSection.fromJson(
        json['biometric'] as Map<String, dynamic>? ?? const {},
      ),
    );
  }
}

class ConsentTypeStatus {
  const ConsentTypeStatus({
    required this.consentType,
    required this.published,
    required this.satisfied,
    this.currentVersion,
    this.acceptedVersion,
    this.acceptedAt,
    this.contentSource,
  });

  final String consentType;
  final bool published;
  final bool satisfied;
  final String? currentVersion;
  final String? acceptedVersion;
  final DateTime? acceptedAt;
  final String? contentSource;

  factory ConsentTypeStatus.fromJson(Map<String, dynamic> json) {
    return ConsentTypeStatus(
      consentType: json['consent_type'] as String? ?? '',
      published: json['published'] == true,
      satisfied: json['satisfied'] == true,
      currentVersion: json['current_version'] as String?,
      acceptedVersion: json['accepted_version'] as String?,
      acceptedAt: json['accepted_at'] == null
          ? null
          : DateTime.tryParse(json['accepted_at'].toString()),
      contentSource: json['content_source'] as String?,
    );
  }
}

class LegalConsentStatus {
  const LegalConsentStatus({
    required this.businessName,
    required this.businessCode,
    required this.legalPageUrl,
    required this.legalSatisfied,
    required this.biometricSatisfied,
    required this.terms,
    required this.privacy,
    required this.biometric,
    this.consentsCompleted = false,
    this.isDemo = false,
  });

  final String businessName;
  final String businessCode;
  final String legalPageUrl;
  final bool legalSatisfied;
  final bool biometricSatisfied;
  final bool consentsCompleted;
  final bool isDemo;
  final ConsentTypeStatus terms;
  final ConsentTypeStatus privacy;
  final ConsentTypeStatus biometric;

  factory LegalConsentStatus.fromJson(Map<String, dynamic> json) {
    return LegalConsentStatus(
      businessName: json['business_name'] as String? ?? '',
      businessCode: json['business_code'] as String? ?? '',
      legalPageUrl: json['legal_page_url'] as String? ?? '',
      legalSatisfied: json['legal_satisfied'] == true,
      biometricSatisfied: json['biometric_satisfied'] == true,
      consentsCompleted: json['consents_completed'] == true,
      isDemo: json['is_demo'] == true,
      terms: ConsentTypeStatus.fromJson(
        json['terms'] as Map<String, dynamic>? ?? const {},
      ),
      privacy: ConsentTypeStatus.fromJson(
        json['privacy'] as Map<String, dynamic>? ?? const {},
      ),
      biometric: ConsentTypeStatus.fromJson(
        json['biometric'] as Map<String, dynamic>? ?? const {},
      ),
    );
  }
}

class ConsentDocumentItem {
  const ConsentDocumentItem({
    required this.id,
    required this.title,
    required this.position,
    required this.url,
    required this.required,
    required this.accepted,
    this.fileUrl,
    this.webPath,
    this.consentType,
    this.version,
  });

  final String id;
  final String title;
  final int position;
  final String url;
  final bool required;
  final bool accepted;
  final String? fileUrl;
  final String? webPath;
  final String? consentType;
  final String? version;

  factory ConsentDocumentItem.fromJson(Map<String, dynamic> json) {
    return ConsentDocumentItem(
      id: json['id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      position: (json['position'] as num?)?.toInt() ?? 0,
      url: json['url'] as String? ?? '',
      required: json['required'] == true,
      accepted: json['accepted'] == true,
      fileUrl: json['file_url'] as String?,
      webPath: json['web_path'] as String?,
      consentType: json['consent_type'] as String?,
      version: json['version'] as String?,
    );
  }
}

class EmployeeConsents {
  const EmployeeConsents({
    required this.items,
    required this.consentsCompleted,
  });

  final List<ConsentDocumentItem> items;
  final bool consentsCompleted;

  factory EmployeeConsents.fromJson(Map<String, dynamic> json) {
    final raw = json['items'];
    return EmployeeConsents(
      items: raw is List
          ? raw
              .whereType<Map<String, dynamic>>()
              .map(ConsentDocumentItem.fromJson)
              .toList()
          : const [],
      consentsCompleted: json['consents_completed'] == true,
    );
  }
}

List<ConsentDocumentItem> pendingRequiredConsents(
  List<ConsentDocumentItem> items,
) {
  final pending = items.where((item) => item.required && !item.accepted).toList();
  pending.sort((a, b) => a.position.compareTo(b.position));
  return pending;
}

